miden-lib/asm/miden/sat/internal/prologue.masm

```diff
@@ -679,78 +679,99 @@ end
 proc.process_consumed_notes_data
     # load the consumed notes data onto the advice stack
     exec.layout::get_nullifier_com adv.push_mapval dropw
+    # => [...]
 
     # read the number of consumed notes from the advice provider
     adv_push.1
+    # => [num_notes, ...]
 
     # store the number of consumed notes
     dup exec.layout::set_total_num_consumed_notes
+    # => [num_notes, ...]
 
-    # assert the number of consumed notes is within limits
-    dup exec.constants::get_max_num_consumed_notes lte assert
+    # assert the number of consumed notes is within limits; since max number of consumed notes is
+    # smaller than 2^32, we can use a more efficient u32 comparison
+    dup exec.constants::get_max_num_consumed_notes u32assert2 u32lte assert
```

Comment:
why does `u32assert2` only need 1 cycle whereas `u32assert` needs 3?

Reply:
`u32assert2` is a native VM operation while `u32assert` is emulated as `push.0 u32assert2 drop`. The main reason for making `u32assert2` operation native was that we frequently need to assert that top 2 values on the stack are `u32` values (more frequently than we need to assert that top value is `u32`).